packages/@o3r-training/tools/src/utility/web-container.ts

```diff
@@ -0,0 +1,85 @@
+import type { BufferEncoding, DirEnt, FileSystemTree } from '@webcontainer/api';
+
+/**
+ * Function to read a directory in the file system
+ */
+export type ReadDirFn = (root: string, options: {
+  encoding?: BufferEncoding | null | undefined;
+  recursive?: boolean | undefined;
+  withFileTypes?: boolean;
+}) => Promise<DirEnt<string>[]>;
+
+/**
+ * Function to read a file in the file system
+ */
+export type ReadFileFn = (root: string, encoding?: BufferEncoding | null | undefined) => Promise<string>;
+
+/**
+ * File system operations
+ */
+export type FileSystem = {
+  /** Reads a given directory and returns its files and directories */
+  readdir: ReadDirFn;
+  /** Reads a given file */
+  readFile: ReadFileFn;
+};
+
+/**
+ * Reads a given directory and returns the files and their corresponding contents
+ * @param entry
+ * @param path
+ * @param fileSystem
+ * @param exclusionList
+ */
+const readDirRec = async (entry: DirEnt<string>, path: string, fileSystem: FileSystem, exclusionList: string[] = []):
+  Promise<FileSystemTree | undefined> => {
+  const entryPath = path + '/' + entry.name;
+  if (exclusionList.includes(entry.name)) {
+    return;
+  }
+  if (entry.isDirectory()) {
+    const dirEntries = await fileSystem.readdir(entryPath, { encoding: 'utf-8', withFileTypes: true });
+    const tree = { [entry.name]: { directory: {} } };
+    for (const subEntry of dirEntries) {
+      tree[entry.name].directory = {
+        ...tree[entry.name].directory,
+        ...(await readDirRec(subEntry, entryPath, fileSystem, exclusionList))
+      };
+    }
+    return tree;
+  }
+  return {
+    [entry.name]: {
+      file: { contents: await fileSystem.readFile(entryPath, 'utf-8') }
+    }
+  };
+};
+
+/**
+ * Get the file system tree from the given file system taking into account the list of files/directories to exclude
+ * @param files
+ * @param fileSystem
+ * @param exclusionList
+ */
+export const getFilesTree = async (files: {isDir: boolean; path: string}[], fileSystem: FileSystem, exclusionList: string[] = []) => {
+  const tree: FileSystemTree = {};
+  for (const {isDir, path} of files) {
+    if (isDir) {
+      const dirEntries = await fileSystem.readdir(path, { encoding: 'utf-8', withFileTypes: true });
+      for (const entry of dirEntries) {
+        const subTree = await readDirRec(entry, path, fileSystem, exclusionList);
+        if (subTree) {
+          tree[entry.name] = subTree[entry.name];
+        }
+      }
+    } else {
+      const name = path.match(/([^\\/]+)$/g)?.[0];
```

Comment:
why not use `path.basename` ?
If we can't then we should add a comment to explain why I think, and it will be easier to understand what this RegExp does

Reply:
This method is both executed in node and webcontainer filesystem (hence why we need to pass the readdir and readfile functions) and unfortunately path.basename does not seem to existe in the WebContainer.fs :(

Reply:
We could add a method next to readdir and readfile

Reply:
I mean that this is not implemented and we would have to do this anyway :)